### DISCUSSION TITLE
Multiple Model Outputs

### DISCUSSION THREAD

**Discussion by CalmLogarithm** [stellargraph/stellargraph#1826] [General]
I've built a model that GraphSAGE and link_inference to implement a multi-class classifier. I wanted to add a second, parallel output layer. It's fairly simple to do with the Keras model. The hangup I'm having is that currently my test/train pairs come from EdgeSplitter -> GraphSAGELinkGenerator -> flow. It's not obvious to me that I can use the flow method to produce a flow with multiple different targets that map to the same IDs. Is there a blessed way to accomplish this?

**Comment by CalmLogarithm** [comment]
I may have a workable solution. Unfortunately, it didn't provide me the magical training improvements that I imagined. But I wanted to document it for other people. I created a custom sequence that combined two sequences, one for each output. First, if you are shuffling your training sequences then you have to give them the same seed value otherwise their orders will be different.

```
train_flow1 = train_gen.flow(edgelist_train, labels_train1, shuffle=True, seed=42)
train_flow2 = train_gen.flow(edgelist_train, labels_train2, shuffle=True, seed=42)
```

(Note, this example will generate the same sequence order every time since I'm passing the constant value, 42. Generate a random integer and use it for both seed values for production use.)

Now the sequence itself wraps the two component sequences:

```
class CombinedSequence(tf.keras.utils.Sequence):
   def __init__(self, seq_a, seq_b):
      if len(seq_a) != len(seq_b):
         raise TypeError("All component sequences must be the same length")
      self.seq_a = seq_a
      self.seq_b = seq_b
   def __len__(self):
      return len(self.seq_a)
   def __getitem__(self, idx):
      x, y0 = self.seq_a[idx]
      _, y1 = self.seq_b[idx]
      return x, [y0, y1]
   def on_epoch_end(self):
      self.seq_a.on_epoch_end()
      self.seq_b.on_epoch_end()
```

**Comment by adocherty** [comment]
Thanks for posting a solution to your query!  I think that another way to do this is to create an output with another dimension using, for example, the `output_dim` argument of the  `link_classification` helper function (noting that the default `edge_embedding_method="ip"` doesn't support this), for example: 

```
prediction = link_classification(
    output_dim=2, output_act="relu", edge_embedding_method="mul"
)(x_out)
```

Note the final dimension in the output shape of this layer is two:
```
> prediction.shape
TensorShape([None, 2])
````

Alternatively, you can stack the outputs of two Keras layers so that the final dimension is 2 (one dimension for each of your outputs).

Then, the edge labels are created as a two dimensional array of shape `(N, 2)` for example assuming the two one-dimensional vectors `labels_train1` and `labels_train2` we can stack them together as follows:

```
edge_labels_train = np.column_stack([labels_train1, labels_train2])
```

The same can be done for the test labels. Everything else in the code should follow, and note that the predictions will have two values for each node, for example:
```
> model.predict(test_flow).shape
(1084, 2)
```

I hope that this helps!
